webpack/src/_parameters.txt
@param options
@text Option Settings

@param 3dMenu
@text 3D Options Menu
@desc Whether 3D options will be in a submenu, regular options menu, or disabled.
@parent options
@type Select
@option SUBMENU
@option ENABLE
@option DISABLE
@default ENABLE


@param renderDistOptionName
@text Render Distance Option Name
@desc symbol name: mv3d-renderDist
@parent options
@type Text
@default Render Distance

@param renderDist
@text Render Distance Default
@desc The maximum distance that can be rendered by the camera.
@parent renderDistOptionName
@type Number
@default 25
@min 0

@param renderDistOption
@text Render Distance Option
@desc Should Render Distance appear on options menu?
@parent renderDistOptionName
@type Boolean
@default true

@param renderDistMin
@text Render Distance Min
@parent renderDistOptionName
@type Number
@default 10
@min 0

@param renderDistMax
@text Render Distance Max
@parent renderDistOptionName
@type Number
@default 100
@min 0


#param mipmapOptionName
#text Mipmapping Option Name
#desc symbol name: mv3d-mipmap
#parent options
#type Text
#default Mipmapping

#param mipmap
#text Mipmapping Default
#parent mipmapOptionName
#type Boolean
#default false

#param mipmapOption
#text Mipmapping Option
#desc Should Mipmapping appear on options menu?
#parent mipmapOptionName
#type Boolean
#default true


@param fovOptionName
@text FOV Option Name
@desc symbol name: mv3d-fov
@parent options
@type Text
@default FOV

@param fov
@text FOV Default
@parent fovOptionName
@type Number
@default 65
@min 0 @max 180

@param fovOption
@text FOV Option
@desc Should FOV appear on options menu?
@parent fovOptionName
@type Boolean
@default false

@param fovMin
@text FOV Min
@parent fovOptionName
@type Number
@default 50
@min 0 @max 180

@param fovMax
@text FOV Max
@parent fovOptionName
@type Number
@default 100
@min 0 @max 180

@param invertYOptionName
@text Invert Y Axis Option Name
@desc symbol name: mv3d-inverty
@parent options
@type Text
@default Invert Y Axis

@param invertYOption
@text Invert Y Axis Option
@desc Should invert Y axis appear on options menu?
@parent invertYOptionName
@type Boolean
@default true

@param lookSensitivityOptionName
@text Look Sensitivity Option Name
@desc symbol name: mv3d-looksensitivity
@parent options
@type Text
@default Look Sensitivity

@param lookSensitivityOption
@text Look Sensitivity Option
@desc Should look sensitivity appear in options menu
@parent lookSensitivityOptionName
@type Boolean
@default true

@param spacer|graphics @text‏‏‎ ‎@desc ===============================================

@param graphics
@text Graphics

@param antialiasing
@text Antialiasing
@parent graphics
@type Boolean
@default true

@param edgefix
@text Edge Fix
@desc Fixes rendering issues at the edges of tiles.
@parent graphics
@type Number
@decimals 1
@default 0.5

@param alphatest
@text Alpha Cutoff
@desc Pixels with alpha below this value will not be rendered.
@parent graphics
@type Number
@decimals 2
@min 0.01 @max 1
@default 0.51

@param lightLimit
@text Lights Per Mesh
@parent graphics
@type Number
@min 4
@default 8

@param backfaceCulling
@text Backface Culling
@parent graphics
@type Boolean
@default true

@param cameraCollision
@text Camera Collision
@parent graphics
@type Combo
@option Off
@option Type1
@option Type2
@option Type1 Smoothed
@option Type2 Smoothed
@default Type2 Smoothed

@param resScale
@text Resolution Scale
@desc Scale the resolution
@parent graphics
@type Number
@decimals 2
@min 0 @max 1
@default 1

@param spacer|map @text‏‏‎ ‎@desc ===============================================

@param map
@text Map Settings

@param cellSize
@text Cell Size
@desc The size of the chunks the map is divided into.
@parent map
@type Number
@default 10

@param unloadCells
@text Unload Far Cells
@desc Unload cells outside the render distance.
@parent map
@type Boolean
@default false

@param eventsUpdateNear
@text Update All Events in Render Distance
@parent map
@type Boolean
@default true

@param mapDefaults
@text Map Defaults
@parent map
@type Note
@default
"enable(true)\nsun(white)\nambient(default)\nfog(black|20,30)\nceiling(backface:true)"


@param spacer|input @text‏‏‎ ‎@desc ===============================================

@param input
@text Input & Gameplay

@param inputCameraMouse
@text Mouse Camera Control
@parent input
@type Boolean
@default false

@param inputCameraGamepad
@text Gamepad Camera Control
@parent input
@type Boolean
@default true

@param inputGamepadTurnButton
@text Gamepad Turn Button
@parent input
@type Select
@option Bumpers
@option Triggers
@option Unchanged
@default Bumpers

@param WASD
@text WASD
@parent input
@type Boolean
@default true

@param dir8Movement
@text Diagonal Movement
@desc In smart mode, when diagonal movement fails, try moving straight.
@parent input
@type Combo
@option Off
@option Diagonal Basic
@option Diagonal Basic 3D Only
@option Diagonal Smart
@option Diagonal Smart 3D Only
@default Diagonal Smart

@param keyboardPitch
@text Control Pitch
@parent input
@desc Allow player to change pitch with pageup & pagedown.
@type Boolean
@default true

@param keyboardTurn
@text Allow Turning
@parent input
@desc Allows rotating camera with keyboard in 3rd person. Rotating is always allowed in 1st person.
@type Select
@option Turn with Q+E (requires WASD) @value QE
@option Turn with A+D and left+right @value AD
@option disabled
@default QE

@param keyboardStrafe
@text Allow Strafing
@parent input
@desc Allows strafing in 1st person mode.
@type Select
@option Strafe with Q+E (requires WASD) @value QE
@option Strafe with A+D and left+right @value AD
@option disabled
@default QE

@param turnIncrement
@text Turn Increment
@parent input
@desc How many degrees the camera will turn when you press the turn button. Other numbers may also be entered.
@type Combo
@option 90
@option 45
@option SMOOTH
@default 90

@param yawSpeed
@text Turn Speed
@parent input
@desc Speed when turning with keyboard (in degrees per second).
@type Number
@default 180

@param pitchSpeed
@text Pitch Speed
@parent input
@desc Speed when changing pitch with keyboard (in degrees per second).
@type Number
@default 90


@param stairThresh
@text Stair Threshold
@desc If the distance in height between two tiles is less than this, they will be passable.
@parent input
@type Number
@decimals 2
@default 0.1

@param walkOffEdge
@text Walk off Edge
@parent input
@type Boolean
@default false

@param walkOnEvents
@text Walk on Events
@parent input
@type Boolean
@default true

@param gravity
@text Gravity
@desc The speed characters will fall, in tiles per second.
@parent input
@type Number
@decimals 2
@default 8

@param spacer|tileconfig @text‏‏‎ ‎@desc ===============================================

@param tileconfig
@text Tile Config

@param wallHeight
@text Wall Height
@desc The default height for wall tiles
@parent tileconfig
@type Number
@min -9999 @max 9999
@decimals 1
@default 2.0

@param tableHeight
@text Table Height
@desc The default height for table tiles
@parent tileconfig
@type Number
@min -9999 @max 9999
@decimals 2
@default 0.33

@param fringeHeight
@text Fringe Height
@parent tileconfig
@type Number
@min -9999 @max 9999
@decimals 1
@default 2.0

@param ceilingHeight
@text Ceiling Height
@desc Default height of ceiling for maps with ceiling enabled.
@parent tileconfig
@type Number
@min -9999 @max 9999
@decimals 1
@default 2.0

@param layerDist
@text Layering Distance
@desc The distance between tile layers. If this is too small
there may be z-fighting issues. (default: 0.0100)
@parent tileconfig
@type Number
@decimals 4
@default 0.0100

@param animDelay
@text Animation Speed
@desc The number of milliseconds between each frame in tile animations.
@parent tileconfig
@type Number
@default 333

@param regions
@text Regions
@desc use regions to determine tile height.
@parent tileconfig
@type struct<RegionHeight>[]
@default ["{\"regionId\":\"1\",\"conf\":\"height(1)\"}","{\"regionId\":\"2\",\"conf\":\"height(2)\"}","{\"regionId\":\"3\",\"conf\":\"height(3)\"}","{\"regionId\":\"4\",\"conf\":\"height(4)\"}","{\"regionId\":\"5\",\"conf\":\"height(5)\"}","{\"regionId\":\"6\",\"conf\":\"height(6)\"}","{\"regionId\":\"7\",\"conf\":\"height(7)\"}"]

@param ttags
@text Terrain Tags
@desc use terrain tags to determine tile height.
@parent tileconfig
@type struct<TTagHeight>[]
@default ["{\"terrainTag\":\"1\",\"conf\":\"shape(xcross),height(1),fringe(0)\"}","{\"terrainTag\":\"2\",\"conf\":\"shape(fence),height(1)\"}"]

@param spacer|characters @text‏‏‎ ‎@desc ===============================================

@param characters
@text Characters

@param heightTrigger
@text Event Trigger Height
@desc If true, events will need to be at the same elevation as the player to be triggered.
@parent characters
@type Boolean
@default true

@param eventCharDefaults
@text Character Event Settings
@parent characters
@type Note
@default "shadow(0.8,4)\nshape(sprite)\nscale(1)"

@param eventObjDefaults
@text Object Event Settings
@parent characters
@type Note
@default "shadow(0)\nshape(sprite)\nscale(1)"

@param eventTileDefaults
@text Tile Event Settings
@parent characters
@type Note
@default "shadow(0)\nshape(flat)\nscale(1)"

@param eventHeight
@text Event "Above Characters" Default Height
@parent characters
@type Number
@decimals 1
@default 2.0

@param boatSettings
@text Boat Settings
@parent characters
@type struct<BoatStruct>
@default {"conf":"shadow(0.8,4),shape(sprite),scale(1),bush(false)"}

@param shipSettings
@text Ship Settings
@parent characters
@type struct<BoatStruct>
@default {"conf":"shadow(0.8,4),shape(sprite),scale(1),bush(false)"}

@param airshipSettings
@text Airship Settings
@parent characters
@type struct<AirshipStruct>
@default {"conf":"shadow(1,6),shape(sprite),scale(1),bush(false)","height":"2.0","bushLanding":"false"}

@param allowGlide
@text Allow Glide
@desc If true, collision detection for flying characters will use only current elevation and not target elevation.
@parent characters
@type Boolean
@default true

@param spriteOffset
@text Sprite Offset
@parent characters
@type Number
@min 0 @max 1
@decimals 2
@default 0.9

@param spacer|assets @text‏‏‎ ‎@desc ===============================================

@param assets
@text Assets

@param diagSymbol
@text Diagonal Sprite Symbol
@desc Character sheets with this symbol contain diagonal sprites. Leave blank to use diagonal sprites with all images.
@parent assets
@type Text
@default {d}

@param shadowTexture
@text Shadow Texture
@parent assets
@type file
@dir img/MV3D
@require 1
@default shadow

@param alphaMask
@text Bush Alpha Texture
@parent assets
@type file
@dir img/MV3D
@require 1
@default bushAlpha

@param errorTexture
@text Error Texture
@parent assets
@type file
@dir img/MV3D
@require 1
@default errorTexture

@param requiredImages
@text Other Required Images
@desc Specify additional images here to prevent them from being excluded during deployment.
@parent assets
@type file[]
@dir img/MV3D
@require 1

webpack/src/parameters.js
import mv3d from './mv3d.js';
import { booleanString,falseString, makeColor, degtorad, tileSize, assign } from './util.js';
import { ORTHOGRAPHIC_CAMERA } from './mod_babylon.js';
import { Attribute } from './attributes.js';

let pluginName = 'mv3d';
if(!PluginManager._scripts.includes("mv3d")){
	if(PluginManager._scripts.includes("mv3d-babylon")){ pluginName='mv3d-babylon'; }
}

const parameters = PluginManager.parameters(pluginName);
export default parameters;

function parameter(name,dfault,type){
	return name in parameters ? (type?type(parameters[name]):parameters[name]) : dfault;
}

Object.assign(mv3d,{
	enumOptionModes:{
		DISABLE: 0,
		ENABLE: 1,
		SUBMENU: 2,
	}
});

function tryParseString(s){
	try{
		return JSON.parse(s);
	}catch(err){
		return s;
	}
}

assign(mv3d,{
	CAMERA_MODE:"PERSPECTIVE",
	ORTHOGRAPHIC_DIST:100,
	MV3D_FOLDER:"img/MV3D",

	ANIM_DELAY:Number(parameters.animDelay),
	ALPHA_CUTOFF:Math.max(0.01,parameters.alphatest),

	EDGE_FIX: Number(parameters.edgefix)*tileSize()/48,
	ANTIALIASING: booleanString(parameters.antialiasing),
	FOV:Number(parameters.fov),
	RES_SCALE: parameter('resScale',1,Number)||1,

	WALL_HEIGHT:Number(parameters.wallHeight),
	TABLE_HEIGHT:Number(parameters.tableHeight),
	FRINGE_HEIGHT:Number(parameters.fringeHeight),
	CEILING_HEIGHT:Number(parameters.ceilingHeight),
	LAYER_DIST:Number(parameters.layerDist),

	//ENABLED_DEFAULT: booleanString(parameters.enabledDefault),
	EVENTS_UPDATE_NEAR: booleanString(parameters.eventsUpdateNear),

	UNLOAD_CELLS: booleanString(parameters.unloadCells),
	CELL_SIZE: Number(parameters.cellSize),
	RENDER_DIST: Number(parameters.renderDist),
	MIPMAP:booleanString(parameters.mipmap),

	MAP_DEFAULTS: parameter('mapDefaults',"",JSON.parse),

	get renderDist(){ return Math.min(this.RENDER_DIST, mv3d.blendFogFar.currentValue()+7.5); },

	OPTION_MIPMAP:booleanString(parameters.mipmapOption),
	OPTION_NAME_MIPMAP: parameter('mipmapOptionName',"Mipmapping",String),
	OPTION_RENDER_DIST: parameter('renderDistOption',true,booleanString),
	OPTION_NAME_RENDER_DIST: parameter('renderDistOptionName',"Render Distance",String),
	OPTION_RENDER_DIST_MIN: parameter('renderDistMin',10,Number),
	OPTION_RENDER_DIST_MAX: parameter('renderDistMax',100,Number),
	OPTION_FOV: parameter('fovOption',false,booleanString),
	OPTION_NAME_FOV: parameter('fovOptionName',"FOV",String),
	OPTION_FOV_MIN: parameter('fovMin',50,Number),
	OPTION_FOV_MAX: parameter('fovMax',100,Number),
	OPTION_INVERTY: parameter('invertYOption',true,booleanString),
	OPTION_NAME_INVERTY: parameter('invertYOptionName',"Invert Y Axis", String),
	OPTION_LOOKSENSITIVITY: parameter('lookSensitivityOption',true,booleanString),
	OPTION_NAME_LOOKSENSITIVITY: parameter('lookSensitivityOptionName',"Look Sensitivity", String),

	STAIR_THRESH: Number(parameters.stairThresh),
	WALK_OFF_EDGE:booleanString(parameters.walkOffEdge),
	WALK_ON_EVENTS:booleanString(parameters.walkOnEvents),
	GRAVITY:Number(parameters.gravity),

	FOG_COLOR: 0,
	FOG_NEAR: 20,
	FOG_FAR: 30, 
	//AMBIENT_COLOR: makeColor(parameters.ambientColor).toNumber(),
	get AMBIENT_COLOR(){ return mv3d.featureEnabled('dynamicShadows')?0x888888:0xffffff; },

	LIGHT_LIMIT: parameter('lightLimit',8,n=>{n=Number(n);return isFinite(n)?n:8;}),
	LIGHT_HEIGHT: 0.5,
	LAMP_HEIGHT: 0.5,
	FLASHLIGHT_HEIGHT: 0.25,
	LIGHT_DECAY: 1,
	LIGHT_DIST: 3,
	LIGHT_ANGLE: 60,
	FLASHLIGHT_EXTRA_ANGLE: 10,
	FLASHLIGHT_INTENSITY_MULTIPLIER: 2,

	REGION_DATA:{},
	_REGION_DATA:{},
	_REGION_DATA_MAP:{},
	TTAG_DATA:{},

	EVENT_HEIGHT:Number(parameters.eventHeight),
	//VEHICLE_BUSH:booleanString(parameters.vehicleBush),
	BOAT_SETTINGS:JSON.parse(parameters.boatSettings),
	SHIP_SETTINGS:JSON.parse(parameters.shipSettings),
	AIRSHIP_SETTINGS:JSON.parse(parameters.airshipSettings),

	EVENT_CHAR_SETTINGS: parameter('eventCharDefaults',"",tryParseString),
	EVENT_OBJ_SETTINGS: parameter('eventObjDefaults',"",tryParseString),
	EVENT_TILE_SETTINGS: parameter('eventTileDefaults',"",tryParseString),

	allowGlide: new Attribute('allowGlide',booleanString(parameters.allowGlide),booleanString),
	get ALLOW_GLIDE(){ return this.allowGlide; },

	SPRITE_OFFSET:Number(parameters.spriteOffset)/2,

	ENABLE_3D_OPTIONS:mv3d.enumOptionModes[parameters['3dMenu'].toUpperCase()],

	TEXTURE_SHADOW: parameters.shadowTexture||'shadow',
	TEXTURE_BUSHALPHA: parameters.alphaMask||'bushAlpha',
	TEXTURE_ERROR: parameters.errorTexture||'errorTexture',

	diagonalMovement: new Attribute('diagonalMovement',String(parameters.dir8Movement),function(v){
		v=String(v).toUpperCase();
		return {
			enabled:booleanString(v),
			smart:v.includes('SMART'),
			'2D':!v.includes('3D'),
		}
	}),
	get DIR8MOVE(){ return this.diagonalMovement.enabled; },
	get DIR8SMART(){ return this.diagonalMovement.smart; },
	get DIR8_2D(){ return this.diagonalMovement['2D']; },

	turnIncrement: new Attribute('turnIncrement',String(parameters.turnIncrement),v=>Number(v)),
	get TURN_INCREMENT(){ return mv3d.turnIncrement; },
	WASD: booleanString(parameters.WASD),

	inputCameraMouse: new Attribute('inputCameraMouse',String(parameters.inputCameraMouse),booleanString),
	inputCameraGamepad: new Attribute('inputCameraGamepad',String(parameters.inputCameraGamepad),booleanString),
	GAMEPAD_TURN_BUTTON: parameter('inputGamepadTurnButton',0,v=>{return {'U':0,'B':1,'T':2}[v[0]]||0;}),

	KEYBOARD_PITCH: booleanString(parameters.keyboardPitch),
	KEYBOARD_TURN: falseString(parameters.keyboardTurn),
	KEYBOARD_STRAFE: falseString(parameters.keyboardStrafe),

	YAW_SPEED: Number(parameters.yawSpeed)||90,
	PITCH_SPEED: Number(parameters.pitchSpeed)||90,

	TRIGGER_INFINITE: !booleanString(parameters.heightTrigger),

	BACKFACE_CULLING: parameter('backfaceCulling',true,booleanString),
	cameraCollision: new Attribute('cameraCollision',String(parameters.cameraCollision),function(v){
		if(typeof v === 'string'){
			const values=v.split(' '); v=values[0];
			const ret = {type:!booleanString(v)?0:Number((v.match(/\d+/)||'1')[0])};
			for (let v of values)if(v.toUpperCase().includes("SMOOTH")){ ret.smooth=true; }
			return ret;
		}else if(typeof v == 'object'){
			return Object.assign(this.get(),v);
		}
		const ret = this.get(); ret.type=Number(v); return ret;
	}),

	DIAG_SYMBOL: parameter('diagSymbol','{d}',String),

	setupParameters(){
		this.REGION_DATA=new Proxy(this._REGION_DATA,{
			get:(target,key)=>{
				if(key in this._REGION_DATA_MAP){ return this._REGION_DATA_MAP[key]; }
				if(key in this._REGION_DATA){ return this._REGION_DATA[key]; }
			},
			set:(target,key,value)=>{
				target[key]=value;
			},
			has:(target,key)=>{
				return key in this._REGION_DATA_MAP || key in this._REGION_DATA;
			},
		});
		for (let entry of JSON.parse(parameters.regions)){
			entry=JSON.parse(entry);
			const regionData = this.readConfigurationFunctions(entry.conf,this.tilesetConfigurationFunctions)
			this._REGION_DATA[entry.regionId]=regionData;
			
		}
		for (const id in this._REGION_DATA){
			this.applyTextureConfigs(this._REGION_DATA[id],'B',0,0);
			this.collapseCeilingOffsets(this._REGION_DATA[id]);
		}
		for (let entry of JSON.parse(parameters.ttags)){
			entry=JSON.parse(entry);
			this.TTAG_DATA[entry.terrainTag]=this.readConfigurationFunctions(entry.conf,this.tilesetConfigurationFunctions);
		}
		for (const id in this.TTAG_DATA){
			this.applyTextureConfigs(this.TTAG_DATA[id],'B',0,0);
			this.collapseCeilingOffsets(this.TTAG_DATA[id]);
		}

		this.EVENT_CHAR_SETTINGS = this.readConfigurationFunctions(
			this.EVENT_CHAR_SETTINGS,
			this.eventConfigurationFunctions,
		);
		this.EVENT_OBJ_SETTINGS = this.readConfigurationFunctions(
			this.EVENT_OBJ_SETTINGS,
			this.eventConfigurationFunctions,
		);
		this.EVENT_TILE_SETTINGS = this.readConfigurationFunctions(
			this.EVENT_TILE_SETTINGS,
			this.eventConfigurationFunctions,
		);

		this.BOAT_SETTINGS.big=booleanString(this.BOAT_SETTINGS.big);
		this.SHIP_SETTINGS.big=booleanString(this.SHIP_SETTINGS.big);
		this.AIRSHIP_SETTINGS.height=Number(this.AIRSHIP_SETTINGS.height);
		this.AIRSHIP_SETTINGS.big=booleanString(this.AIRSHIP_SETTINGS.big);
		this.AIRSHIP_SETTINGS.bushLanding=booleanString(this.AIRSHIP_SETTINGS.bushLanding);

		this.BOAT_SETTINGS.conf = this.readConfigurationFunctions(
			this.BOAT_SETTINGS.conf,
			this.eventConfigurationFunctions,
		);
		this.SHIP_SETTINGS.conf = this.readConfigurationFunctions(
			this.SHIP_SETTINGS.conf,
			this.eventConfigurationFunctions,
		);
		this.AIRSHIP_SETTINGS.conf = this.readConfigurationFunctions(
			this.AIRSHIP_SETTINGS.conf,
			this.eventConfigurationFunctions,
		);

		//Texture.DEFAULT_ANISOTROPIC_FILTERING_LEVEL=0;

		this.MAP_DEFAULTS = this.readConfigurationFunctions(this.MAP_DEFAULTS,this.mapConfigurationFunctions);
	},

	updateParameters(){
		this.updateRenderDist();
		this.updateFov();
		this.callFeatures('updateParameters');
	},
	updateRenderDist(){
		if(this.camera.mode===ORTHOGRAPHIC_CAMERA){
			this.camera.maxZ=this.renderDist;
			this.camera.minZ=-this.renderDist;
		}else{
			this.camera.maxZ=this.renderDist;
			this.camera.minZ=0.1;
		}
	},
	updateFov(){
		const dist = this.blendCameraDist.currentValue()||0.1;
		const frustrumHeight = this.getFrustrumHeight(dist,degtorad(this.FOV));
		const fov = this.getFovForDist(dist,frustrumHeight/this.blendCameraZoom.currentValue());
		this.camera.fov=fov;
	},
});